modules/tests/shared/src/test/scala/io/circe/CursorSuite.scala

```diff
@@ -0,0 +1,301 @@
+package io.circe
+
+import cats._
+import cats.syntax.all._
+import io.circe.syntax._
+import io.circe.tests.CirceMunitSuite
+import org.scalacheck.Prop
+import org.scalacheck.Prop._
+
+final class CursorSuite extends CirceMunitSuite {
+  import CursorSuite._
+
+  val j1: Json = Json.obj(
+    "a" -> (1 to 5).toList.asJson,
+    "b" -> Map("d" -> List(true, false, true)).asJson,
+    "c" -> Map("e" -> 100.1, "f" -> 200.2).asJson
+  )
+
+  val j2: Json = Json.obj(
+    "a" -> (0 to 5).toList.asJson,
+    "b" -> Map("d" -> List(true, false, true)).asJson,
+    "c" -> Map("e" -> 100.1, "f" -> 200.2).asJson
+  )
+
+  val j3: Json = Json.obj(
+    "a" -> (1 to 5).toList.asJson,
+    "b" -> 10.asJson,
+    "c" -> Map("e" -> 100.1, "f" -> 200.2).asJson
+  )
+
+  val j4: Json = Json.obj(
+    "a" -> (1 to 5).toList.asJson,
+    "c" -> Map("e" -> 100.1, "f" -> 200.2).asJson
+  )
+
+  val cursor: Cursor = Cursor.fromJson(j1)
+
+  property("focus should return the JSON value in a newly created cursor") {
+    forAll { (j: Json) =>
+      Cursor.fromJson(j).focus ?= Some(j)
+    }
+  }
+
+  property("Navigation into an object")(topProp)
+  private lazy val topProp = forAll { (j: Json) =>
+    val c = Cursor.fromJson(j)
+
+    val intoObject = for {
+      keys <- c.keys
+      first <- keys.headOption
+    } yield c.downField(first)
+
+    forAllP(intoObject)(value =>
+      ((value.top.fold(Prop.passed)(_ ?= j)) :| "Should return to top") &&
+        ((value.root.focus.fold(Prop.passed)(_ ?= j)) :| "Should return to root") &&
+        ((value.up.focus.fold(Prop.passed)(_ ?= j)) :| "Up should undo this operation")
+    )
+  }
+
+  property("Navigation into an array") {
+    forAll { (j: Json) =>
+      val c: Cursor = Cursor.fromJson(j).downArray
```

Comment:
Would these tests (navigate into object/array) be better if the `forAll` generated objects and arrays rather than generic json?